src/Paprika/Chain/PooledSpanDictionary.cs

```diff
@@ -0,0 +1,220 @@
+﻿using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Paprika.Store;
+
+namespace Paprika.Chain;
+
+/// <summary>
+/// The pooled span dictionary uses memory from <see cref="BufferPool"/> to store large chunks of memory.
+/// </summary>
+/// <remarks>
+/// This component hacks into the dictionary to provide the underlying keys by implementing the custom equality comparer.
+/// It allows to store in the underlying dict just simple entries and not pass the reference in there. To get the reference,
+/// the <see cref="Equals"/> gets the underlying <see cref="Span{Byte}"/> from pages;
+/// </remarks>
+public class PooledSpanDictionary : IEqualityComparer<PooledSpanDictionary.KeySpan>, IDisposable
+{
+    private const int BufferSize = BufferPool.BufferSize;
+    private readonly BufferPool _pool;
+    private readonly Dictionary<KeySpan, ValueSpan> _dict;
+    private readonly List<Page> _pages = new();
+
+    private Page _current;
+    private int _position;
+
+    private readonly byte[] _key = new byte[1024];
```

Comment:
Why 1024? Are we sure all keys and values will fit into that memory? And for keys only seems excessive? Shouldn't be `MaxKeySize`?

Reply:
It's a dummy constant. Will provide better.